Fingerprint Dialog & MAM

lovetox opened the pull request:
- Fingerprint Dialog from ChatWindow
- Config Dialog from Plugin Window
- MAM Message handling
- Trust of Fingerprint is checked before sending and receiving Messages
- Better Warning about what the current state of the Plugin is, and if a Message arrived unencrypted through MAM

Commit:
Initial commit Config Dialog

Commit:
show current OMEMO encryption state in chat Window

Commit:
Handle MAM Message received Event

Commit:
DB Migration and fixed douple entrys in DB

Commit:
Removing "GetDeviceKey" Button & fetch keys automatically

Commit:
Finterprint Dialog Button in ChatWindow added

Commit:
Trust is checked before Message is sent & small other fixes

Commit:
Warn user when new Fingerprints are in DB and refreshAuthLockSymbol()

Commit:
Made one Verify Button instead of Trust/Untrust Button

Commit:
handle Message received from an Untrusted Fingerprint

Commit:
Catch an error when opening config dialog

If the config dialog was opend before any chat window ui_list[] was empty and a key error was thrown

akallabeth commented:
@lovetox using your pull and the UI improvements are impressive! The fingerprint dialog does not properly resize to multiple fingerprints though.

kalkin commented on __init__.py:
```diff
@@ -128,42 +131,102 @@ def _compute_caps_hash(self):
+            omemo_enabled = state.encryption.is_active(jid)
+
+            if omemo_enabled:
+                msg.msgtxt = '**Unencrypted** ' + msg.msgtxt
```
We shouldn't do that. I don't trust GTK, Gajim & Python Unicode and character handling. I could imagine people trying to send weird Unicode chars to hide the `**Unencrypted**` prefix. You should warn in the ui with the green text. Have a look at ui.py line 120 `plain_warning`.

kalkin replied:
Same applies for the non mam message handling

lovetox replied:
i thought about that, but GAJIM is not capable for a good solution right now
this is basically a ugly hack.

the problem is not with messages that arrive while the UI is open, we could even just show the plain warning over the gajim api.

the problem lies with MAM.
when you receive MAM Messages there is no UI (Chat Window) open, so you cant use the warning over the gajim api.

But in my opinion security wise we should inform the user that unencrypted messages arrived through MAM.

sadly Gajim has no API whatsoever to change what comes into the Messagelog DB, cause my idea would be that we add a column there with "encrypted" true or false. thats not possible at the moment.

so the only way is to edit the message i receive via MAM when no UI is open at the time. its ugly but i see no other option.

Also unencrypted only will be added if the omemo encryption was active with that contact, otherwise the user expects unencrypted traffic and i dont edit the message.

Also the plain warning IS shown when the UI is open, i didnt change that, the **unencrypted** is just an addition.

suggestion how to solve this are welcome

kalkin replied:
> the problem is not with messages that arrive while the UI is open, we could even just show the plain warning over the gajim api.
> 
> the problem lies with MAM.

Ahh, I see. ☹

> But in my opinion security wise we should inform the user that unencrypted messages arrived through MAM.

True that.

> sadly Gajim has no API whatsoever to change what comes into the Messagelog DB, cause my idea would be that we add a column there with "encrypted" true or false. thats not possible at the moment.

Have a look at gajim src code in `src/common/optparser.py` there you can add an migration to the logs and unread(?) table. I'm pretty sure @yann-l will accept a patch for that. @tmolitor could reuse it when he will write the crypto api for Gajim.

lovetox replied:
Thats not enough, even if we migrate the database, MAM sends messages directly to the logs, if you login.
even if i have a database that gives me the correct information, for example that the message was unencrypted, the only place i could show that is in the history window.

and to my knowledge its at the moment not possible to hook into the history window and make new fields for various encryption states. this is of course what should be in a future gajim version, and im looking into it with @tmolitor as soon as i have the feeling that omemo is in a good state.

to resolve this problem, the commit has MAM Message handling in it, which is needed.
if this is important to you i can take the 4-5 lines where the unencrypted gets added out of the commit, so that the rest can pass

kalkin replied:
After some discussion between @lovetox @iNPUTmice and me, we agreed to add it as is

kalkin commented on omemo/encryption.py:
```diff
@@ -92,4 +92,20 @@ def migrate(dbConn):
+                            recipient_id, public_key
+                            );
+                        """
+
```
Don't we need to add an Index to prevent duplicate entries in the future?

lovetox replied:
there cant be any duplicates in the future once this query was run, cause i revised the function who writes into that table to check for jid AND key, before writing.
the function is saveIdentity()

i hope i understood the question right

kalkin replied:
> i revised the function who writes into that table to check for jid AND key, before writing.
> the function is saveIdentity()

This is good, but still add an `UNIQUE` index to the sqlite migration. This make the code less errorprone, if i.e: `saveIdentity()` is changed. Also it follows the logic that the db is responsible for storing and maintaining data.

kalkin commented:
@lovetox Thank you very much for your pull requests. You did an amazing job. I already cherry picked a few atomic commits (1b927a4, fc2b712, 767d8da). I added some notes to clarify some things.

kalkin commented:
Also relevant #38

Commit:
Make small design changes to the Fingerprint UI

Commit:
Add Clear Devices function to the Config UI

- Adds a page where all currently published Device IDs are shown
- Small design changes to the whole Config UI

lovetox commented on omemo/encryption.py:
```diff
@@ -91,5 +91,35 @@ def migrate(dbConn):
+
+    return dbConn
+
+    if user_version(dbConn) < 3:
```
UNIQUE Index is now included

lovetox commented on omemo/encryption.py:
```diff
@@ -91,5 +91,35 @@ def migrate(dbConn):
+                                      %s
+                                      PRAGMA user_version=3;
+                                      END TRANSACTION;
+                                  """ % (add_index))
```
UNIQUE INDEX is now included

lovetox commented on omemo/encryption.py:
```diff
@@ -91,5 +91,34 @@ def migrate(dbConn):
+                                 %s
+                                 PRAGMA user_version=3;
+                                 END TRANSACTION;
+                             """ % (add_index))
```
UNIQUE INDEX now added

kalkin replied:
You can just do it in the same TRANSACTION.

Commit:
Change all arrows to "=>"

Arrows could not be displayed in a Windows Log file

Commit:
Catch an error when the UI is not present

If the UI (Chat Window) is not open a key error is thrown

kalkin commented on __init__.py:
```diff
@@ -164,11 +164,24 @@ def message_received(self, msg):
+                    gui = self.ui_list[account].get(jid, None)
+                    if gui and gui.encryption_active():
+                        gui.plain_warning()
+                except:
```
What kind of exception do we expect here? Catching a general `Exception` is overkill

kalkin commented on __init__.py:
```diff
@@ -253,22 +314,23 @@ def connect_ui(self, chat_control):
+            log.warn(account_name + " ⇒ Missing keys for " + contact_jid + ": " +
                      str(result))
-        return result
+            log.warn('query keys now ...')
```
I think we should change both of the logs to `info`. A missing key is not that exceptional and is part of the normal workflow now

kalkin commented on __init__.py:
```diff
@@ -367,21 +428,10 @@ def session_from_prekey_bundle(self, account_name, state, stanza,
-        if account in self.ui_list:
-            if recipient_id in self.ui_list[account]:
-                self.ui_list[account][recipient_id].update_prekeys()
+            log.warn(recipient_id + ' => session created')
```
`info` should be enough

lovetox replied:
this whole section gets removed in one of the next commits

kalkin commented on omemo/liteidentitykeystore.py:
```diff
@@ -67,10 +69,70 @@ def storeLocalData(self, registrationId, identityKeyPair):
+        return result
+
+    def getFingerprints(self, jid):
+        q = "SELECT _id, recipient_id, public_key, trust FROM identities " + \
```
Please do proper quoting instead of concating strings.

kalkin commented on omemo/state.py:
```diff
@@ -232,7 +235,11 @@ def create_msg(self, from_jid, jid, plaintext):
         for rid, cipher in session_ciphers.items():
             try:
-                encrypted_keys[rid] = cipher.encrypt(key).serialize()
+                if self.trust[self.isTrusted(cipher)] is True:
```
`self.trust[self.isTrusted(cipher)] is True` ? Why not just:

``` python
if self.trust[self.isTrusted(cipher)]:
    …
```

kalkin replied:
Ahh now I understand why you are doing `self.trust[self.isTrusted(cipher)] is True`.

Instead of using a dictionary for `trust` use constants:

``` python
TRUST_NOT_SET  = None
TRUST_FALSE  = 0
TRUST_TRUE = 1
TRUST_UNDECIDED  = 2

if self.getTrustState(cipher) == TRUST_TRUE:
   …
```

Also you should probably replace `"Not Set"` with "`undecided`", or do you have a special purpose for the "Not Set" state?

kalkin commented on omemo/state.py:
```diff
@@ -232,7 +235,11 @@ def create_msg(self, from_jid, jid, plaintext):
+                if self.trust[self.isTrusted(cipher)] is True:
+                    encrypted_keys[rid] = cipher.encrypt(key).serialize()
+                else:
+                    log.warn('Skipped Device because Trust is: ' +
```
You should change it to `debug` or you can change it to `info` but only print it if state is `UNDECIDED`

kalkin commented on omemo/state.py:
```diff
@@ -326,13 +342,22 @@ def get_session_cipher(self, jid, device_id):
+            return key
+        else:
+            raise Exception("Received PreKeyWhisperMessage from Untrusted Fingerprint!")
+            return
```
No need for return after raise

kalkin commented on ui.py:
```diff
@@ -127,5 +122,328 @@ def plain_warning(self):
+        if self.state.store.identityKeyStore.getUndecidedFingerprints(self.contact.jid):
+            msg = "You received a new Fingerprint. " + \
+                  "Until you make a trust decision you can only " + \
+                  "receive encrypted Messages from that Device."
```
You can do just:

``` python
msg = "asdasd " \
      "asdasd "
```

No need for `+`

Commit:
Not allow to send msg if Fingerprint is untrusted

Only if at least one Fingerprint is trusted a message goes out
otherwise a error message is printed

cleaned up SQL querys

removed unnecessary return statements

cippaciong commented:
@kalkin 
I've been using the plugin from lovetox repo for some days on multilple devices by now and it worked really well. Considering that in these days we've told different people to try lovetox version, both on issues on github and in gajim conference, wouldn't be better to integrate these changes in the main repo (on a testing branch in case you don't want to break master) so that people can test them with less efforts and then fix the smaller problems you are discussing right now in further pull requests?
Just my opinion..

lovetox commented:
@cippaciong i fix this little things, over the next 2 days and then we can move forward :)

Commit:
Small code formatting change

Commit:
Better error handling & change log level

kalkin commented:
@cippaciong I don't see a big difference in using a different branch or using @lovetox repo directly.

Also using lovetox branch is **unsupported** and should only done by people who now what they are doing. I.e. the code created db scheme `user_version` 2 and 3, now it only creates 2. So all people who tested lovetox repo will have issues, if we release a db change to `user_version` 3 at some day.

kalkin commented:
@vanitasvitae https://github.com/omemo/gajim-omemo/issues/63#issuecomment-225188107
> @lovetox (Sorry, there is no issue section in your fork, so I don't know how to reach you) I tested your fork across multiple servers. It still seems not to work in all cases. Fingerprints are verified at all ends and omemo checkbox is activated, but sent messages are unencrypted...

vanitasvitae commented:
I started gajim with `gajim -l gajim.plugin_system.omemo=DEBUG` to see if there is some suspicious output, but when I send messages from gajim there is no output at all and when I receive messages there is debug output including 

> Decrypted msg=>'message'

Commit:
Make fingerprint trust code more readable

Commit:
Change level of log message

cippaciong commented:
> [...] I.e. the code created db scheme user_version 2 and 3, now it only creates 2. So all people who tested lovetox repo will have issues, if we release a db change to user_version 3 at some day.
@kalkin Should we set it back to 2 then?

kalkin commented:
@cippaciong probably

kalkin commented:
@cippaciong *Correction* No need. We will leave it as it is.

kalkin commented:
@lovetox Merged. Thank you for your great work
